vpdq/cpp/hashing/filehasher.cpp

```diff
@@ -62,322 +109,412 @@ static void saveFrameToFile(AVFrame* frame, const char* filename) {
-  int width = downsampleWidth;
-  int height = downsampleHeight;
-  if (width == 0) {
-    width = codecParameters->width;
-  }
-  if (height == 0) {
-    height = codecParameters->height;
-  }
+    if (codec->capabilities & AV_CODEC_CAP_FRAME_THREADS) {
+      codecContext->thread_type = FF_THREAD_FRAME;
+    } else if (codec->capabilities & AV_CODEC_CAP_SLICE_THREADS) {
+      codecContext->thread_type = FF_THREAD_SLICE;
+    } else {
+      codecContext->thread_count = 1;
+    }
 
-  if (width == 0 || height == 0) {
-    std::cerr << "Width or height equals 0" << std::endl;
-    avformat_close_input(&formatContext);
-    return false;
-  }
+    // Open the codec context
+    if (avcodec_open2(codecContext.get(), codec, nullptr) < 0) {
+      throw std::runtime_error("Cannot open video codec context");
+    }
 
-  std::unique_ptr<vpdq::hashing::AbstractFrameBufferHasher> phasher =
-      vpdq::hashing::FrameBufferHasherFactory::createFrameHasher(height, width);
-  if (phasher == nullptr) {
-    std::cerr << "phasher allocation failed" << std::endl;
-    avformat_close_input(&formatContext);
-    return false;
-  }
+    // Get the framerate
+    AVRational avframeRate =
+        formatContext->streams[videoStreamIndex]->avg_frame_rate;
 
-  // Find the video decoder
-  const AVCodec* codec = avcodec_find_decoder(codecParameters->codec_id);
-  if (!codec) {
-    std::cerr << "Codec decoder not found" << std::endl;
-    avformat_close_input(&formatContext);
-    return false;
+    // if avg_frame_rate is 0, fall back to r_frame_rate which is the
+    // lowest framerate with which all timestamps can be represented accurately
+    if (avframeRate.num == 0 || avframeRate.den == 0) {
+      avframeRate = formatContext->streams[videoStreamIndex]->r_frame_rate;
+    }
+
+    frameRate = static_cast<double>(avframeRate.num) /
+        static_cast<double>(avframeRate.den);
+    if (frameRate == 0) {
+      throw std::runtime_error("Video framerate is zero");
+    }
   }
 
-  // Create the codec context
-  AVCodecContext* codecContext = avcodec_alloc_context3(codec);
-  if (avcodec_parameters_to_context(codecContext, codecParameters) < 0) {
-    std::cerr << "Cannot copy codec parameters to context" << std::endl;
-    avformat_close_input(&formatContext);
-    return false;
+  void createSwsContext() {
+    swsContext = SwsContextPtr(sws_getContext(
+        codecContext->width,
+        codecContext->height,
+        codecContext->pix_fmt,
+        width,
+        height,
+        PIXEL_FORMAT,
+        DOWNSAMPLE_METHOD,
+        nullptr,
+        nullptr,
+        nullptr));
+
+    if (swsContext.get() == nullptr) {
+      throw std::runtime_error("Cannot create sws context");
+    }
   }
+};
+
+class vpdqHasher {
+ public:
+  struct FatFrame {
+    AVFramePtr frame;
+    int64_t frameNumber;
+  };
+
+  std::condition_variable queue_condition;
+  std::mutex queue_mutex;
+  std::queue<FatFrame> hash_queue;
+  std::atomic<bool> done_hashing{false};
```

Comment:
You are using std::atomic for this variable, but everywhere you are using it, you are holding the queue_mutex, so you don't need an std::atomic.

Furthermore, I think you are relying on mutex to make sure the condition variable reads correctly (it's set before notification), so I think better to convert to just a bool to prevent a future author thinking its threadsafe to use as is.

There was a pattern I used in a previous project that wrapped variables in a mutex which only made them accessible if the mutex was held, which make things like this much less error prone, but I can't find it in the core CPP libraries, so it must have been bespoke.